cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,387 @@
+package db
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"time"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/support/db"
+	"github.com/stellar/go/support/errors"
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/go/xdr"
+
+	"github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/ledgerbucketwindow"
+)
+
+const (
+	transactionTableName = "transactions"
+)
+
+type Transaction struct {
+	Result           []byte   // XDR encoded xdr.TransactionResult
+	Meta             []byte   // XDR encoded xdr.TransactionMeta
+	Envelope         []byte   // XDR encoded xdr.TransactionEnvelope
+	Events           [][]byte // XDR encoded xdr.DiagnosticEvent
+	FeeBump          bool
+	ApplicationOrder int32
+	Successful       bool
+	Ledger           ledgerbucketwindow.LedgerInfo
+}
+
+// TransactionWriter is used during ingestion
+type TransactionWriter interface {
+	InsertTransactions(lcm xdr.LedgerCloseMeta) error
+	RegisterMetrics(ingest, insert, count prometheus.Observer)
+}
+
+// TransactionReader is used to serve requests and just returns a cloned,
+// read-only DB session to perform actual DB reads on.
+type TransactionReader interface {
+	NewTx(ctx context.Context) (TransactionReaderTx, error)
+	GetLedgerRange(ctx context.Context) ledgerbucketwindow.LedgerRange
+}
+
+// TransactionReaderTx provides all of the public ways to read from the DB.
+// Note that `Done()` *MUST* be called to clean things up.
+type TransactionReaderTx interface {
+	GetTransaction(log *log.Entry, hash xdr.Hash) (Transaction, bool, ledgerbucketwindow.LedgerRange)
+	GetLedgerRange() ledgerbucketwindow.LedgerRange
+	Done() error
+}
+
+type transactionHandler struct {
+	db         db.SessionInterface
+	stmtCache  *sq.StmtCache
+	passphrase string
+
+	ingestMetric, insertMetric, countMetric prometheus.Observer
+}
+
+type transactionReaderTx struct {
+	ctx        context.Context
+	db         db.SessionInterface
+	passphrase string
+}
+
+func NewTransactionReader(db db.SessionInterface, passphrase string) TransactionReader {
+	return &transactionHandler{db: db, passphrase: passphrase}
+}
+
+func NewTransactionWriter(db db.SessionInterface, stmtCache *sq.StmtCache, passphrase string) TransactionWriter {
+	return &transactionHandler{db: db, stmtCache: stmtCache, passphrase: passphrase}
+}
+
+func (txn *transactionHandler) InsertTransactions(lcm xdr.LedgerCloseMeta) error {
+	start := time.Now()
+	txCount := lcm.CountTransactions()
+	L := log.
```

Comment:
the pattern soroban rpc has adapted is having a logger be passed in to each component. you can checkout how it's done in cmd/soroban-rpc/internal/ingest/service.go . Also, I noticed that we don't log anything in any of the other db components so I think it makes sense to avoid logging here so that we can have a consistent style.

Reply:
> you can checkout how it's done in cmd/soroban-rpc/internal/ingest/service.go

Can you elaborate? I see the initial [logger creation](https://github.com/stellar/soroban-rpc/blob/main/cmd/soroban-rpc/internal/ingest/service.go#L82) and its usage in the `service` itself (e.g. [here](https://github.com/stellar/soroban-rpc/blob/main/cmd/soroban-rpc/internal/ingest/service.go#L267)), but I don't see it passed along to any subservice.

> Also, I noticed that we don't log anything in any of the other db components so I think it makes sense to avoid logging here so that we can have a consistent style.

I'd argue it makes more sense to start logging in the other DB components, especially since we're moving towards a DB-only architecture and we want the same logging/metrics/visibility that we used to have in the in-memory components.

Reply:
you can pass in a logger when creating the DB instance instead of using the global logger. this thread is also relevant:

https://github.com/stellar/soroban-rpc/pull/141/files/67d5ed9e6f2c7378d78b4a89223a3e415d6a6aa5#r1578459290

Reply:
Yep you're right, I think I was just plumbing it incorrectly when I tried it before. Fixed up in f187405!